Replaced `declare_tailcall_if` with logic in the loader

dylandreimerink opened the pull request:
This PR replaces the `declare_tailcall_if` macro with loader logic in Go. Before this we had the `declare_tailcall_if` and `invoke_tailcall_if`. These were used together to determine if a function should be inlined or be emitted and called as a tailcall. For this to work, both the declare and invoke side needed to have the same condition or risk a missed tailcall.

This change simplifies the datapath code a bit without performance regressions. We now always emit all tailcalls into the ELF, but we use logic in the loader to see if any `invoke_tailcall_if` or any non-conditional `tail_call_static` references a given tailcall. If a tailcall is unreachable it will not be loaded, saving us load time.

This is part of a larger movement from compile time logic to load time logic to enable clang-free in the future, and to make the datapath code easier today.

```release-note
Replaced `declare_tailcall_if` with logic in the loader
```

dylandreimerink committed:
pkg/bpf: Implement unreachable tail call pruning

This commit implements unreachable tail call pruning. When loading a
collection we check if a tail call is reachable. If not, we remove the
tail call from the collection. This saves us from having to load the
tail call program into the kernel.

Previously, we would conditionally not include tail calls in the
collection with pre-processor directives. Now that we do it in the
loader, we can remove the pre-processor directives.

Signed-off-by: Dylan Reimerink <dylan.reimerink@isovalent.com>

dylandreimerink committed:
bpf: Remove `declare_tailcall_if`

Remove `declare_tailcall_if`, so we always emit the tailcall programs
into the ELF. The followup commit will implement pruning logic based
on the actual usage of the tail calls. This means that we will only
need the `invoke_tailcall_if` without the need to keep both the
declaration and invocation in sync.

Signed-off-by: Dylan Reimerink <dylan.reimerink@isovalent.com>

dylandreimerink committed:
pkg/bpf: Add test for removeUnreachableTailcalls

This commit adds a test to verify the behavior of the dead tail call
pruning. It consists of 5 tail calls, of which 2 are unreachable. The
test asserts that only the unreachable tail calls are removed from the
spec.

Signed-off-by: Dylan Reimerink <dylan.reimerink@isovalent.com>

lmb commented on pkg/bpf/collection.go:
```diff
@@ -41,6 +45,110 @@ func LoadCollectionSpec(path string) (*ebpf.CollectionSpec, error) {
+		}
+
+		var id, slot uint32
+		if _, err := fmt.Sscanf(prog.SectionName, "%d/%v", &id, &slot); err != nil {
```
I think this can scan less than you think. It probably makes sense to check the first return value.

> It returns the number of items successfully parsed.

dylandreimerink replied:
I don't understand. You want to check that the first return value == 2? I assumed this would work given I copied it from `iproute2Compat`.

lmb replied:
Never mind, I was thinking of C `sscanf` semantics: https://go.dev/play/p/eAbk0sJrNeI

lmb commented on pkg/bpf/collection.go:
```diff
@@ -41,6 +45,110 @@ func LoadCollectionSpec(path string) (*ebpf.CollectionSpec, error) {
+			continue
+		}
+
+		// Any programs not in the tail call map can be ignored (always load them).
```
`always load them` confuses me, shouldn't we add them to `entrypoints` then?

dylandreimerink replied:
Hmm. I added this as guard condition I didn't expect to be actually used. But I guess that if someone starts using a new type of section name, and that new program uses tail calls, then we would want to treat it as an entrypoint. That would make this code more future proof. :+1:

lmb commented on pkg/bpf/collection.go:
```diff
@@ -41,6 +45,110 @@ func LoadCollectionSpec(path string) (*ebpf.CollectionSpec, error) {
+	visit := func(prog *ebpf.ProgramSpec, tailcalls map[uint32]*TailCall) {
+		for i := 0; i < len(prog.Instructions); i++ {
+			inst := prog.Instructions[i]
+			if !(inst.IsBuiltinCall() && inst.Constant == int64(asm.FnTailCall)) {
```
Nit: I'd find `!Builtin || Constant != TailCall` easier to read.

lmb commented on pkg/bpf/collection.go:
```diff
@@ -41,6 +45,110 @@ func LoadCollectionSpec(path string) (*ebpf.CollectionSpec, error) {
+				continue
+			}
+
+			// For all tail calls where the index is static we use the `tail_call_static` function.
```
Does it make sense to document in the C for `tail_call_static` that this function relies on the specific asm?

lmb replied:
Maybe even add sig-loader to CODEOWNERS for that file.

dylandreimerink replied:
Yea, I think you are right.

lmb commented on pkg/bpf/collection.go:
```diff
@@ -41,6 +45,110 @@ func LoadCollectionSpec(path string) (*ebpf.CollectionSpec, error) {
+			// tail call instruction is a `mov r3, <index>` instruction. The IMM value of this
+			// instruction is used as the index for the tail call.
+			prev := prog.Instructions[i-1]
+			if !(prev.OpCode.Class() == asm.ALU64Class && prev.Dst == asm.R3) {
```
Just checking for class doesn't mean it's a `mov`, no?

dylandreimerink replied:
:facepalm:

lmb commented on pkg/bpf/collection.go:
```diff
@@ -41,6 +45,110 @@ func LoadCollectionSpec(path string) (*ebpf.CollectionSpec, error) {
+
+			tc := tailcalls[uint32(prev.Constant)]
+			if tc == nil {
+				// This tends to happen due to the custom tail calls map which also uses
```
I think you can retrieve the name of the map from the `ldwimm64` used to load the map fd, no? Then we could actually check here that this references the correct map. As is is right now I'm a bit worried that this is too loose.

dylandreimerink replied:
:thinking: , yea, it should have map reference. I will try this, see if I can make it work.

dylandreimerink replied:
So yes and no. It isn't as easy as just looking at 2 instructions above the tail call. That instruction is always `Mov R2, Rx`. And then its up to the compiler to pick a register and make sure it contains the fd. So far it always appears to go:
```
Mov Rx, <map fd>
Mov R1, Rctx
Mov R2, Rx
Mov R3, <index>
Call tail_call
```

So I wrote some code that looks for this. However, if the compiler ever decides to use a stack slot for this instead of re-getting the map fd we are strewed. While I think this is unlikely, if I am wrong, we will get random missing tail-calls.

For now I want to see if this works in CI. If not then we have to revert to not knowing, or perhaps change `tail_call_static` to emit `Mov R2, <map fd>` instead (if that is possible). 

Last option is that if we are unable to find the `Mov Rx, <map fd>`, that we assume the worst case and mark the index as referenced, that way we load a dead program instead of having missing tail calls.

lmb replied:
Weird, why does the compiler emit the extra `mov Rx`?

Anyways, can we somehow make the compiler emit a relocation for the `mov r2`? Maybe we can use the `CONFIG` macro hack?

```
asm volatile("%[out] = __config_" #name " ll" : [out]"=r"(out));
```

(Might also be able to use the ll suffix to skip the separate load?)

gentoo-root replied:
> Weird, why does the compiler emit the extra mov Rx?

I think it's because we have `"r2 = %[map]\n\t"`, `[map]"r"(map)` in tail_call_static, so the compiler substitutes another register (`"r"`) to be `%[map]`. Or did you mean anything else?

dylandreimerink replied:
I managed to get this to work :partying_face:

lmb commented on pkg/bpf/collection.go:
```diff
@@ -41,6 +45,110 @@ func LoadCollectionSpec(path string) (*ebpf.CollectionSpec, error) {
+	tailcalls := make(map[uint32]*TailCall)
+
+	for _, prog := range spec.Programs {
+		if prog.SectionName == "tc" || prog.SectionName == "xdp" {
```
Why not check `prog.Type` instead? This is brittle because we also allow these to be prefixes like `tc/foobar`.

Similarly, only programs where `prog.Type == Unspecified` can be tail calls due to our weird section naming. Maybe it's good to not rely on that though.

ti-mo replied:
@lmb The kernel doesn't accept Unspecified progs. These are fixed up in `classifyProgramTypes`.

@dylandreimerink Could we remove some of this logic by running tail call pruning after `iproute2Compat` and `classifyProgramTypes`? After those functions, `cilium_calls` will have its `MapSpec.Contents` correctly populated, and I think we can limit the pruning algorithm to progs appearing in cilium_calls only. For example, entrypoint progs will never appear in cilium_calls.

dylandreimerink replied:
Why not check prog.Type instead? Not sure these are set since the `classifyProgramTypes` program that explicitly sets these runs after this one. (I think it might work now even without the classifyProgramTypes)

In any case, I suggest we remove this specific logic, and let the if not a `x/y` section treat as entrypoint logic catch these, that should work regardless of whatever changes may happen in the future.

lmb commented on pkg/bpf/testdata/Makefile:
Does this need to be hooked into CI somehow to make sure the file is regenerated?

dylandreimerink committed:
bpf: Modify tail_call_static to emit better parseable assembly

Before this change the tail_call_static function would emit the
following instructions to perform a tailcall:

```
Mov R1, Rctx
Mov R2, Rmap_ptr
Mov R3, <slot>
Call TailCall
```

Since the second instruction is always a Register to Register move, we
would have to backtrack to find the actual map which is being used.

These changes makes it so the following instructions are emitted:

```
Mov R1, Rctx
Mov R2, 0 ll <calls_map>
Mov R3, <slot>
Call TailCall
```

By always using a double word immediate, with a relocation entry on the
Mov R2 instruction it is much easier to find the actual map which is
being used. As a side effect, we usually eliminate an extra instruction
clang was otherwise forced to emit.

Signed-off-by: Dylan Reimerink <dylan.reimerink@isovalent.com>

lmb commented on CODEOWNERS:
```diff
@@ -270,6 +270,7 @@ Makefile* @cilium/build
 /bpf/lib/auth.h @cilium/sig-datapath @cilium/sig-servicemesh
 /bpf/lib/encrypt.h @cilium/ipsec
 /bpf/lib/wireguard.h @cilium/wireguard @cilium/sig-datapath
+/bpf/include/bpf/tailcall.h @cilium/sig-datapath @cilium/loader
```
Maybe just loader?

lmb commented on pkg/bpf/collection.go:
```diff
@@ -41,6 +48,135 @@ func LoadCollectionSpec(path string) (*ebpf.CollectionSpec, error) {
 }
 
+func RemoveUnreachableTailcalls(spec *ebpf.CollectionSpec) error {
+	return removeUnreachableTailcalls(spec)
```
Why the unexported function?

dylandreimerink replied:
Ah, that was temporary while testing from outside the package, it can go now.

lmb commented on pkg/bpf/collection.go:
```diff
@@ -41,6 +48,135 @@ func LoadCollectionSpec(path string) (*ebpf.CollectionSpec, error) {
+		var id, slot uint32
+		// Consider any program that doesn't follow the tailcall naming convention
+		// x/y to be an entrypoint.
+		if _, err := fmt.Sscanf(prog.SectionName, "%d/%v", &id, &slot); err != nil {
```
Follow up question: why not `%d` twice?

Also: would something break if we gave this a prefix like `static_tail_call/%d` and then got rid of the `cilium_calls_map` logic below?

dylandreimerink replied:
Comments on the original say its so we accept hex encoding as well (`0xffff`) for example, but I don't thing this is used outside of the policy template when not replaced. I will test if removing it breaks anything.

lmb commented on pkg/bpf/collection.go:
```diff
@@ -41,6 +48,135 @@ func LoadCollectionSpec(path string) (*ebpf.CollectionSpec, error) {
+		}
+
+		const (
+			cilium_calls_map = 2
```
How is this magic number going to stay in sync?

ti-mo replied:
We can consider removing the map ID system by using section names like `cilium_calls/23`, but that would require some work on the ELF rewriter since they're actually called `cilium_calls_01234` in the ELF.

If/when we remove ELF rewriting, my intention was to fix the map name to `cilium_calls` in the ELF and change the `MapSpec.Name` at runtime instead. This results in `CollectionSpec.Maps["cilium_calls"]` but `MapSpec.Name = "cilium_calls_01234"`, so the map can always be retrieved using a stable key.

dylandreimerink replied:
We had a meeting since, agreed that removing the ID system is the goal long term, but since that will take a bit of time to move forward with the magic IDs for now.

lmb commented on pkg/bpf/collection.go:
```diff
@@ -41,6 +48,135 @@ func LoadCollectionSpec(path string) (*ebpf.CollectionSpec, error) {
+			continue
+		}
+
+		tailcalls[slot] = &TailCall{
```
Paranoid mode: check for duplicate slots? I guess that isn't possible since section names can't be the same?

dylandreimerink replied:
Hmm, we should not have duplicates since section names are unique, but with the scanf that accepts multiple encodings it can technically happen (`2/1` and `2/0x1`). Added extra guard rails can't hurt.

lmb commented on pkg/bpf/collection.go:
```diff
@@ -41,6 +48,135 @@ func LoadCollectionSpec(path string) (*ebpf.CollectionSpec, error) {
+
+	// Allow test_cilium_calls_[0-9]+, cilium_calls_[0-9]+, cilium_calls_hostns_[0-9]+, and cilium_calls_netdev_[0-9]+
+	// Compile in func and not in global scope to allow for the names to be updated.
+	callsMapRegex := regexp.MustCompile(fmt.Sprintf(
```
Ummm, I'm not sure how to feel about this. First, panicking is bad. Second, stuffing the strings into a regex verbatim makes my spidey sense tingle.

I'm guessing we can't change the name of these maps without things breaking terribly?

dylandreimerink replied:
Yes, I hate this as well. The `test_` variant really skewed me while trying to get CI green. I currently don't really have any other way to identify static from dynamic maps other than name. And as you can see we are horribly inconsistent with our map naming.

I think we can perhaps use the legacy iproute ID, but that only lasts until we remove that in favor of sane logic.

I honestly don't know why the name for this map has to be so dynamic, perhaps we should just hardcode the name of the `CALLS_MAP` to something and then enforce in CI that it doesn't get touched. I think we can work out in the loader which programs do and don't share a calls maps (which seems to be the purpose of the different names at the moment)

dylandreimerink replied:
I replaced this regex with a string.Contains using the the `callsmap.MapName` gloabal variable. That should have the same effect and be slightly less fishy. In the long term the idea is to make the name of this map hardcoded or to move towards defining the tail calls statically in the map definition. 

For now this is the best we can do.

lmb commented on pkg/bpf/collection.go:
```diff
@@ -41,6 +48,135 @@ func LoadCollectionSpec(path string) (*ebpf.CollectionSpec, error) {
+			}
+
+			// We should at least have 3 instructions before the tail call.
+			if i-3 <= 0 {
```
Or: start the loop at `i := 3; i < len(prog.Instructions)...`.

lmb commented on pkg/bpf/collection.go:
```diff
@@ -41,6 +48,135 @@ func LoadCollectionSpec(path string) (*ebpf.CollectionSpec, error) {
+			if tc == nil {
+				// This tends to happen due to the custom tail calls map which also uses
+				// the static tailcall calling.
+				continue
```
So we still have the continue, can't we return an error now?

lmb commented on pkg/bpf/collection.go:
```diff
@@ -41,6 +48,135 @@ func LoadCollectionSpec(path string) (*ebpf.CollectionSpec, error) {
+				continue
+			}
+
+			ref := movR2.Reference()
```
What if: instead of hardcoding the list of allowable maps, add a prefix in the inline asm. Then make sure that the reference here has that prefix, and otherwise `continue`. If the prefix is present we set `referenced = true` and do `ins.SetReference(strings.TrimPrefix("inline_prefix", ...))`.

That way we can avoid knowledge of what the map name should be here, the only thing that matters is that the tail call is coming in via the tail_call_static macro. The downside is that now we need to modify the eBPF to make it load :(

dylandreimerink replied:
As discussed offline, lets keep this as is for now and improve the situation with BTF on calls maps and a better map pinning scheme.

learnitall reviewed (approved):
Looks good for docs!

aditighag reviewed (approved):
+1 to Loren'z comment for the CODEOWNERS change, approving for `contributing`.

lmb commented on pkg/bpf/collection.go:
```diff
@@ -41,6 +48,130 @@ func LoadCollectionSpec(path string) (*ebpf.CollectionSpec, error) {
+		}
+
+		const (
+			cilium_calls_map = 2
```
Can you add a warning to `bpf/lib/ids.h` and also make that owned by `sig-loader`?

ti-mo commented on pkg/bpf/testdata/unreachable-tailcall.c:
```diff
@@ -0,0 +1,144 @@
+};
+
+
+struct bpf_elf_map {
```
Can we import these from actual headers used in the datapath instead? (also the other macros below)

dylandreimerink replied:
Hmm, let me try. I think so if we add something like a `-I $(PROJECT_ROOT)/bpf/lib` to the clang invocation

ti-mo commented on pkg/bpf/collection.go:
```diff
@@ -41,6 +48,130 @@ func LoadCollectionSpec(path string) (*ebpf.CollectionSpec, error) {
+		var id, slot uint32
+		// Consider any program that doesn't follow the tailcall naming convention
+		// x/y to be an entrypoint.
+		if _, err := fmt.Sscanf(prog.SectionName, "%d/%v", &id, &slot); err != nil {
```
We could probably do
```suggestion
		if _, err := fmt.Sscanf(prog.SectionName, "%d/%v", &id, &slot); err != nil || id != cilium_calls_map {
```
and eliminate the second loop.

dylandreimerink replied:
Sorry, I don't get it. That would save an additional if statement, but I don't see how it eliminates one of the loops

ti-mo replied:
Sorry, yes, I mean the `if id != cilium_calls_map { ...` statement below.

dylandreimerink committed:
bpf: Fix missing tail calls

The changes to the dead tail call elimination revealed 2 cases of
missing tail calls.

First is to do with NAT46x64 logic where there
still existed a call path from the IPv4 logic which would attempt to
tail call into IPv6 to recirculate the packet, even when the IPv6 tail
call wasn't compiled in.

The second was that when XDP offloaded, the IPv6 logic would tail call
into a ICMP6 tail call which is only compiled in for TC programs.

This commit fixes both possible missing tail calls.

Signed-off-by: Dylan Reimerink <dylan.reimerink@isovalent.com>

ti-mo commented on pkg/bpf/collection.go:
```diff
@@ -41,6 +48,138 @@ func LoadCollectionSpec(path string) (*ebpf.CollectionSpec, error) {
+
+			tc := tailcalls[uint32(movIdx.Constant)]
+			if tc == nil {
+				return fmt.Errorf("tail call to unknown index '%d' @ %d, potential missed tailcall", movIdx.Constant, realOff)
```
Not sure if insn index is valuable to the reader, I'd remove it.

```suggestion
				return fmt.Errorf("missed tail call to program at cilium_calls slot %d", movIdx.Constant)
```

ti-mo commented on pkg/bpf/collection.go:
```diff
@@ -41,6 +48,138 @@ func LoadCollectionSpec(path string) (*ebpf.CollectionSpec, error) {
+
+			// Find the tail call instruction.
+			inst := prog.Instructions[i]
+			realOff++
```
Nit: this is typically called the 'raw' instruction count.

ti-mo commented on pkg/bpf/collection.go:
```diff
@@ -41,6 +48,138 @@ func LoadCollectionSpec(path string) (*ebpf.CollectionSpec, error) {
+			// Find the tail call instruction.
+			inst := prog.Instructions[i]
+			realOff++
+			if inst.OpCode.IsDWordLoad() {
```
We have `asm.OpCode.rawInstructions()` for this, but obviously not exported. See my suggestion for using `insns.Iterate()` above.

ti-mo commented on pkg/bpf/collection.go:
```diff
@@ -41,6 +48,138 @@ func LoadCollectionSpec(path string) (*ebpf.CollectionSpec, error) {
+		log.Debugf("visiting section '%s' / prog '%s'", prog.SectionName, prog.Name)
+
+		realOff := 2
+		for i := 3; i < len(prog.Instructions); i++ {
```
Note that we have `asm.Instructions.Iterate()` that gives us a `RawInstructionOffset` out of the box. You'd have to explicitly skip the first 3 insns, but IMO the ergonomics of the iterator make up for that.

ti-mo commented on pkg/bpf/collection.go:
```diff
@@ -41,6 +48,138 @@ func LoadCollectionSpec(path string) (*ebpf.CollectionSpec, error) {
+	visit := func(prog *ebpf.ProgramSpec, tailcalls map[uint32]*TailCall) error {
+		log.Debugf("visiting section '%s' / prog '%s'", prog.SectionName, prog.Name)
+
+		realOff := 2
```
Looks like this is kept for debugging purposes only. I'm questioning the value of keeping this around, since it's still impossible to debug this by only looking at the logs.

So, instead of echoing the raw insn offset, could we print `inst`, `movIdx` and `movR2` in error/debug messages instead?

ti-mo commented on pkg/bpf/collection.go:
```diff
@@ -41,6 +48,138 @@ func LoadCollectionSpec(path string) (*ebpf.CollectionSpec, error) {
+	}
+
+	// Keep visiting tailcalls until no more are discovered.
+outer:
```
It's quite hard to understand what this is supposed to do and how the visiting algorithm works. Adding some comments could improve the situation, but ideally we would remove 1 or 2 levels of indentation here.

ti-mo reviewed (changes requested):
Thanks, this is awesome! :partying_face: 

Ultra-nit: it's a little uncommon to use `@` in error/debug strings, it makes my brain try to parse an email address. :grin: Better to just write 'at' or make it a more natural sentence, e.g. `"ignoring tail call into map '%s' during pruning: map is not cilium_calls", ref`.

I think we only need the following debug messages:
- not a calls map, skipping
- unreferenced, deleting

Reason being that these two mean false negatives and false positives respectively. I'd also explicitly mention the section and program name in each message, since output will appear out of order when tailing multiple Cilium pods at a time, and the `visiting section` message kind of loses its purpose.

ti-mo reviewed (approved):
Awesome, thanks for addressing the feedback!

One nit discussed offline: `if (((char *)data)[0] == 0x01) {` could be replaced with a volatile const instead to remove some irrelevant code from the C test.

julianwiedmann commented on bpf/lib/l3.h:
```diff
@@ -28,15 +28,16 @@
 	ret = ipv6_dec_hoplimit(ctx, l3_off);
 	if (IS_ERR(ret)) {
+#ifndef SKIP_ICMPV6_HOPLIMIT_HANDLING
 		if (ret == DROP_TTL_EXCEEDED)
 			return icmp6_send_time_exceeded(ctx, l3_off, direction);
-
+#endif
```
Oh nice catch, thank you! I think we'll need to wire that up properly at some point for XDP.

dylandreimerink replied:
That is what this is doing. The only place where we ever declare this macro is in bpf_xdp.c. https://github.com/cilium/cilium/blob/main/bpf/bpf_xdp.c#L22

julianwiedmann replied:
I meant that one day we should actually implement the hoplimit handling for XDP :).

julianwiedmann commented on bpf/bpf_host.c:
```diff
@@ -595,13 +595,13 @@ handle_ipv4(struct __ctx_buff *ctx, __u32 secctx __maybe_unused,
 			int ret = nodeport_lb4(ctx, ip4, ETH_HLEN, secctx, ext_err, &is_dsr);
-
+#ifdef ENABLE_IPV6
 			if (ret == NAT_46X64_RECIRC) {
 				ctx_store_meta(ctx, CB_SRC_LABEL, secctx);
 				return tail_call_internal(ctx, CILIUM_CALL_IPV6_FROM_NETDEV,
 							  ext_err);
 			}
-
+#endif
```
Curious, why didn't we catch the same case in `bpf_xdp`?

(I would hope that the agent enforces `IPv6=true` when using the NAT46x64 gateway, so this is probably harmless for real deployments)

dylandreimerink replied:
> Curious, why didn't we catch the same case in bpf_xdp?

It could just be that we are not testing an instance where we have IPv4=true, IPv6=false, NodeportAcceleration=true ?
